File: android-app/app/src/main/java/com/securityradio/ptt/device/ImbeVoiceCodec.kt
package com.securityradio.ptt.device

/**
 * IMBE encoder + decoder, wrapping [P25ImbeNative] in the [VoiceEncoder] /
 * [VoiceDecoder] interfaces so it slots into [VoiceCodecRegistry] alongside
 * Codec2 and Opus. Wire format is unchanged: 2-byte magic (0xF5 0xAB) +
 * 11-byte 88-bit IMBE codeword = 13 bytes total per 20 ms frame.
 *
 * The encoder downsamples the uniform 16 kHz input to 8 kHz internally
 * (matching the historical `downsampleAvg16kToImbe` path). The decoder
 * emits 8 kHz samples; the transport layer upsamples + post-processes via
 * the agency's [PostDecodeChain] before playback.
 */

class ImbeEncoder : VoiceEncoder {
    override val codec: VoiceCodec = VoiceCodec.IMBE
    override val isReady: Boolean get() = P25ImbeNative.isAvailable

    override fun encodeFrame(pcm16kLe640: ByteArray): ByteArray? {
        if (!P25ImbeNative.isAvailable) return null
        if (pcm16kLe640.size < P25ImbeNative.Frames.PCM_16K_FRAME_BYTES) return null
        val imbeIn = P25ImbeNative.Frames.downsampleAvg16kToImbe(pcm16kLe640)
        val codeword = P25ImbeNative.encodeFrame(imbeIn) ?: return null
        val packet = ByteArray(2 + codeword.size)
        packet[0] = codec.magic0
        packet[1] = codec.magic1
        System.arraycopy(codeword, 0, packet, 2, codeword.size)
        return packet
    }
}

class ImbeDecoder : VoiceDecoder {
    override val codec: VoiceCodec = VoiceCodec.IMBE
    override val isReady: Boolean get() = P25ImbeNative.isAvailable
    override val nativeSampleRate: Int = 8000

    override fun decodeFrame(framedBytes: ByteArray): ShortArray? {
        if (!P25ImbeNative.isAvailable) return null
        if (framedBytes.size != 13) return null
        if (framedBytes[0] != codec.magic0 || framedBytes[1] != codec.magic1) return null
        val codeword = framedBytes.copyOfRange(2, 13)
        return P25ImbeNative.decodeCodeword11(codeword)
    }
}

File: ios-app/SafeTMobile/Audio/ImbeVoiceCodec.swift
import Foundation

/// IMBE encoder + decoder, wrapping `P25ImbeNative` in the `VoiceEncoder` /
/// `VoiceDecoder` protocols so it slots into `VoiceCodecRegistry` alongside
/// Codec2 and Opus. Wire format is unchanged: 2-byte magic (0xF5 0xAB) +
/// 11-byte 88-bit IMBE codeword = 13 bytes total per 20 ms frame.

final class ImbeEncoder: VoiceEncoder {
    let codec: VoiceCodec = .imbe
    var isReady: Bool { P25ImbeNative.isAvailable }

    func encodeFrame(_ pcm16kLe640: Data) -> Data? {
        guard isReady else { return nil }
        guard pcm16kLe640.count >= P25ImbeNative.Frames.pcm16kFrameBytes else { return nil }
        guard let imbeIn = P25ImbeNative.Frames.downsampleAvg16kToImbe(frame16k: pcm16kLe640),
              let codeword = P25ImbeNative.encodeFrame(samples8k160: imbeIn) else { return nil }
        var packet = Data(capacity: 2 + codeword.count)
        packet.append(codec.magic0)
        packet.append(codec.magic1)
        packet.append(codeword)
        return packet
    }
}

final class ImbeDecoder: VoiceDecoder {
    let codec: VoiceCodec = .imbe
    var isReady: Bool { P25ImbeNative.isAvailable }
    let nativeSampleRate: Int = 8000

    func decodeFrame(_ framedBytes: Data) -> [Int16]? {
        guard isReady else { return nil }
        guard framedBytes.count == 13 else { return nil }
        let firstByte = framedBytes[framedBytes.startIndex]
        let secondByte = framedBytes[framedBytes.startIndex + 1]
        guard firstByte == codec.magic0, secondByte == codec.magic1 else { return nil }
        let codeword = framedBytes.subdata(in: framedBytes.startIndex + 2..<framedBytes.startIndex + 13)
        return P25ImbeNative.decodeCodeword11(codeword)
    }
}
